feat(aws-ecs): Adds an addSecret() method for a ContainerDefinition

danwiltshire opened the pull request:
I'm using Aspects to configure already-instantiated resources in CDK. This PR will enable adding a Secret to a ContainerDefinition, similar to addEnvironment().

Had to remove the `readonly` modifier to allow the `addSecret()` method to set or push secrets. Let me know if there's a better way of handling this.

Added an integration test, worked as expected in my environment.

closes #18959

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 9a0ae563bcecaff86146d2a6621db4b1c85c51e0
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1db6021f-a7c5-4305-968d-f7dd59be3f8d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 7755250d7032ad2c93a60957f6526960cde8082a
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f7a02e3f-bea1-46f8-b23a-61cfbdab244f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: db7fe407914891bcf99eee1ee20d4fe655cb45fe
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=622ee8ae-1b8e-4af2-a085-0f116e600a90%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 30b6706237ae254144ade219a85e10c0722349e8
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3b94481f-76fd-4493-bcd4-fa952945176c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

robertd commented:
@madeline-k Can you please review this and merge in when time permits? We need this in our current ECS projects & workflows. TIA.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 3f5657b5f1db1ec8e148bccce148515b21d616c2
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1f93949f-2884-41f6-bb7e-c882a2528e53%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f81fbf9518611b72f47889b76cc2e6047eb6c463
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b83ea447-2dc0-4da3-bead-ef93177d30ec%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

madeline-k commented on packages/@aws-cdk/aws-ecs/lib/container-definition.ts:
```diff
@@ -443,7 +443,7 @@ export class ContainerDefinition extends CoreConstruct {
    * stored in Secrets Manager.
    */
-  public readonly referencesSecretJsonField?: boolean;
+  public referencesSecretJsonField?: boolean;
```
This property should only be modifiable from within the class. Can you replace this with a public property getter that references a private property?

danwiltshire replied:
Sure thing, will push a commit.

madeline-k reviewed (dismissed):
Thank you for implementing this feature, @danwiltshire! Nice work. I just have one comment.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a7789c9139b84f29c549fb3a2348531dbc350a71
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bd256780-e7f3-4b24-9096-47802c40028c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 3d3a56a5a4b6cab28dc844065c6318fdfd697764
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7f56f4af-51c0-45b6-be91-fb73556a4b4c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

danwiltshire commented:
@madeline-k Hello, I updated the PR as requested but these tests are now failing.  Do you know how to resolve this?

Thanks

```
API elements with incompatible changes:
err  - PROP @aws-cdk/aws-ecs.ContainerDefinition.referencesSecretJsonField: changed from property to method [changed-kind:@aws-cdk/aws-ecs.ContainerDefinition.referencesSecretJsonField]
err  - PROP @aws-cdk/aws-ecs.FirelensLogRouter.referencesSecretJsonField: changed from property to method [changed-kind:@aws-cdk/aws-ecs.FirelensLogRouter.referencesSecretJsonField]
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: b916590e67cd4df6ff37ed6708234c577729a4e6
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=32c179fe-9f35-40c5-95cd-a052690f5a17%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: b7f84d526906b8006796abb45ac41bf797ab4f25
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c85a391c-f28f-4deb-a81d-9c5ac9ce27a8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in BUILD FAILING for 21 days, and looks abandoned. It will be closed in 10 days if no further commits are pushed to it.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: b720ce03f4fb629d5c8ab0914e72d542bf814760
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=80d314b1-bc9b-4e71-bcc3-3947c69ca374%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 5fcf23bdcbfbea462ef7946168035e5cbee6122c
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=330b2bc6-0a06-4e1e-bbc3-b516639e73f2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

anelson commented:
@madeline-k it looks like this PR has gone stale.  We really need this ability in our project.  Is there anything I could do to get this moving forward?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 199bd32358a28e194f973bf088cc33f41ddfb567
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2d6f9ea4-fe4e-4f2f-91a1-798e9505dd09%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jishi commented on packages/@aws-cdk/aws-ecs/lib/container-definition.ts:
```diff
@@ -606,6 +598,34 @@ export class ContainerDefinition extends CoreConstruct {
+  /**
+   * This method returns true if any Secret in this definition references a JSON field.
+   */
+  public referencesSecretJsonField(): boolean {
```
Maybe u

jishi commented on packages/@aws-cdk/aws-ecs/lib/container-definition.ts:
```diff
@@ -606,6 +598,34 @@ export class ContainerDefinition extends CoreConstruct {
+    } else {
+      return false;
+    }
+  }
```
Maybe use getter syntax to allow it to be accessed as a property is what needs to be changed:

`public referencesSecretJsonField(): boolean {`

That would allow it to be accessed as `obj.referencesSecretJsonField` thus not breaking API spec.
